filtering.py

```diff
@@ -0,0 +1,941 @@
+
+        # loading all models in the case of training
+        if self.training:
+            self.target_models = nn.ModuleList([
+                model.load_model(
+                    target=target,
+                    model_name=self.model_name,
+                    device=self.device
+                ) for target in self.targets
+            ])
+            for target_model in self.target_models:
+                target_model.train()
+        else:
+            self.target_models = None
+
+    def forward(self, audio, rate):
+        """
+        Performing the separation on audio input
+
+        Parameters
+        ----------
+        audio: torch.Tensor [shape=(nb_samples, nb_channels, nb_timesteps)]
+        mixture audio
+
+        rate: int
+        sampling rate of the audio.
+
+        Returns
+        -------
+        estimates: `dict` [`str`, `torch.Tensor`]
+            dictionary of all restimates as performed by the separation model.
+            Note that there may be an additional source in case
+            a residual source is added. If the input audio is 2D, then the
+            first dimension (sample) is removed also from the input.
+
+            Note however, that the last two dimensions are
+            (channel, time steps) in all cases.
+        model_rate: int
+            the new sampling rate, desired by the open-unmix model
+            """
+        if not utils._torchaudio_available():
+            raise Exception('Torchaudio must be available to use the'
+                            'Separator.')
+        import torchaudio
+        from torchaudio.functional import istft
+
+        # shape management
+        remove_samples_dim = False
+        convert_to_numpy = False
+        if self.smart_input_management:
+            shape = torch.tensor(audio.shape)
+            if not isinstance(audio, torch.Tensor):
+                audio = torch.tensor(audio, device=self.device,
+                                     requires_grad=False)
+                convert_to_numpy = True
+            if len(shape) == 1:
+                audio = audio[None, None, ...]
+                remove_samples_dim = True
+            elif len(shape) == 2:
+                if shape.min() <= 2:
+                    audio = audio[None, ...]
+                    remove_samples_dim = True
+                else:
+                    audio = audio[:, None, ...]
+            if audio.shape[1] > audio.shape[2]:
+                audio = audio.transpose(1, 2)
+            if audio.shape[1] > 2:
+                warnings.warn(
+                    'Channel count > 2!. Only the first two channels '
+                    'will be processed!')
+                audio = audio[..., :2]
+        audio = audio.float()
+        if audio.shape[1] == 1:
+            # if we have mono, we duplicate it
+            # as the input of OpenUnmix is always stereo
+            audio = audio.expand(-1, 2, -1)
```

Comment:
I think all these checks should go outside of the forward path as they prevent a potential training to be optimized/vectorized. We need to rely on that users will provide the right input and output. Lets make a check pre-process functions that users can run before using the separator (useful for inference)